httpload/cmd/httpload/main.go

```diff
@@ -0,0 +1,440 @@
+	// copies values from c to out until c is closed, then calls wg.Done.
+	output := func(c <-chan error) {
+		for n := range c {
+			out <- n
+		}
+		wg.Done()
+	}
+	wg.Add(len(cs))
+	for _, c := range cs {
+		go output(c)
+	}
+
+	// Start a goroutine to close out once all the output goroutines are done.
+	// This must start after the wg.Add call.
+	go func() {
+		wg.Wait()
+		close(out)
+	}()
+
+	return out
+}
+
+func (lt *LoadTester) run(config *Config, done chan bool) error {
+	var (
+		ctx    context.Context
+		cancel context.CancelFunc
+	)
+
+	if config.appDuration > 0 {
+		ctx, cancel = context.WithTimeout(context.Background(), config.appDuration)
+	} else {
+		ctx, cancel = context.WithCancel(context.Background())
+	}
+	defer cancel()
+
+	go func() {
+		<-done
+		cancel()
+	}()
+
+	var errcList []<-chan error
+	urlc, errc, err := config.genRequests(ctx)
+	if err != nil {
+		return err
+	}
+	errcList = append(errcList, errc)
+
+	resultc, errc, err := lt.genLoadRequests(ctx, config, urlc)
+	if err != nil {
+		return err
+	}
+	errcList = append(errcList, errc)
+
+	errc, err = lt.genResults(ctx, resultc)
+	if err != nil {
+		return err
+	}
+	errcList = append(errcList, errc)
+
+	err = lt.waitForPipeline(errcList...)
+
+	done <- true
+
+	return err
+}
+
+func (lt *LoadTester) dumpStatuses(w io.Writer) {
+	statuses := make(map[int]int)
+
+	for _, result := range lt.results {
+		statuses[result.status]++
+	}
+
+	keys := make([]int, 0, len(statuses))
+
+	for key := range statuses {
+		keys = append(keys, key)
+	}
+
+	sort.Ints(keys)
+
+	for _, key := range keys {
+		code := float64(key)
+		count := float64(statuses[key])
+		fmt.Fprintf(w, "%s\n", fmt.Sprintf("[%3.f] %12.f response(s)", code, count))
+	}
+}
+
+func (lt *LoadTester) dumpTimings(w io.Writer) {
+	totalTiming := float64(0)
+	maxTiming := float64(0)
+	minTiming := float64(0)
+	averageTiming := float64(0)
+	requestsPerSecond := float64(0)
+
+	n := len(lt.results)
+	noResults := (n == 0)
+
+	if noResults {
+		return
+	}
+
+	timing0 := lt.results[0].timing.Seconds()
+	totalTiming = timing0
+	maxTiming = timing0
+	minTiming = timing0
+
+	for i := 1; i < n; i++ {
+		timing := lt.results[i].timing.Seconds()
+
+		totalTiming += timing
+
+		if timing > maxTiming {
+			maxTiming = timing
+		}
+
+		if timing < minTiming {
+			minTiming = timing
+		}
+	}
+
+	averageTiming = totalTiming / float64(n)
+	requestsPerSecond = float64(1) / averageTiming
+
+	fmt.Fprintf(w, "%s\n", fmt.Sprintf("Total:        %12.4f secs", totalTiming))
+	fmt.Fprintf(w, "%s\n", fmt.Sprintf("Slowest:      %12.4f secs", maxTiming))
+	fmt.Fprintf(w, "%s\n", fmt.Sprintf("Fastest:      %12.4f secs", minTiming))
+	fmt.Fprintf(w, "%s\n", fmt.Sprintf("Average:      %12.4f secs", averageTiming))
+	fmt.Fprintf(w, "%s\n", fmt.Sprintf("Requests/sec: %12.4f", requestsPerSecond))
+}
+
+func setupGracefulShutdown(done chan bool) {
+	signalChannel := make(chan os.Signal, 1)
+
+	signal.Notify(signalChannel, syscall.SIGINT, syscall.SIGKILL)
```

Comment:
Theoretically syscall package should never be used when we're dealing with multiplatform code.
you can use os.Interrupr and os.Kill, see https://godoc.org/os#Signal
At the moment they're aliased of syscall.SIGINT and SIGKILL but who knows it could change in the future